internal/querynodev2/delegator/segment_pruner.go

```diff
@@ -0,0 +1,215 @@
+package delegator
+
+import (
+	"sort"
+	"strconv"
+
+	"github.com/golang/protobuf/proto"
+
+	"github.com/milvus-io/milvus-proto/go-api/v2/commonpb"
+	"github.com/milvus-io/milvus-proto/go-api/v2/schemapb"
+	"github.com/milvus-io/milvus-storage/go/common/log"
+	"github.com/milvus-io/milvus/internal/proto/internalpb"
+	"github.com/milvus-io/milvus/internal/proto/planpb"
+	"github.com/milvus-io/milvus/internal/storage"
+	"github.com/milvus-io/milvus/internal/util/clustering"
+	"github.com/milvus-io/milvus/internal/util/exprutil"
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+	"github.com/milvus-io/milvus/pkg/common"
+	"github.com/milvus-io/milvus/pkg/util/distance"
+	"github.com/milvus-io/milvus/pkg/util/funcutil"
+	"github.com/milvus-io/milvus/pkg/util/merr"
+)
+
+const defaultFilterRatio float64 = 0.5
+
+type PruneInfo struct {
+	filterRatio float64
+}
+
```

Comment:
PruneSegments is a try-best process, we only exclude a candidate segment when it's absolutely sure it will not contain any single value matching the input condition. Otherwise we keep the segment inside the target distribution